plugins/spectate/spectate.cpp

```diff
@@ -0,0 +1,382 @@
+        out.print("auto-unpause: %s\n", unpause_enabled ? "on." : "off.");
+        out.print("auto-disengage: %s\n", disengage_enabled ? "on." : "off.");
+    }
+    saveConfig();
+    return DFHack::CR_OK;
+}
+
+// every tick check whether to decide to follow a dwarf
+void onTick(color_ostream& out, void* ptr) {
+    if (!df::global::ui) return;
+    int32_t tick = df::global::world->frame_counter;
+    if (our_dorf) {
+        if (!Units::isAlive(our_dorf)) {
+            following_dwarf = false;
+            df::global::ui->follow_unit = -1;
+        }
+    }
+    if (!following_dwarf || (focus_jobs_enabled && !job_watched) || (tick - timestamp) > (int32_t) tick_threshold) {
+        std::vector<df::unit*> dwarves;
+        for (auto unit: df::global::world->units.active) {
+            if (!Units::isCitizen(unit)) {
+                continue;
+            }
+            dwarves.push_back(unit);
+        }
+        std::uniform_int_distribution<uint64_t> follow_any(0, dwarves.size() - 1);
+        if (df::global::ui) {
+            // if you're looking at a warning about a local address escaping, it means the unit* from dwarves (which aren't local)
+            our_dorf = dwarves[follow_any(RNG)];
+            df::global::ui->follow_unit = our_dorf->id;
+            job_watched = our_dorf->job.current_job;
+            following_dwarf = true;
+            if (!job_watched) {
+                timestamp = tick;
+            }
+        }
+    }
+}
+
+// every new worked job needs to be considered
+void onJobStart(color_ostream& out, void* job_ptr) {
+    // todo: detect mood jobs
+    int32_t tick = df::global::world->frame_counter;
+    auto job = (df::job*) job_ptr;
+    // don't forget about it
+    int zcount = ++freq[job->pos.z];
+    job_tracker.emplace(job->id);
+    // if we're not doing anything~ then let's pick something
+    if ((focus_jobs_enabled && !job_watched) || (tick - timestamp) > (int32_t) tick_threshold) {
+        following_dwarf = true;
+        // todo: allow the user to configure b, and also revise the math
+        const double b = base;
+        double p = b * ((double) zcount / job_tracker.size());
+        std::bernoulli_distribution follow_job(p);
+        if (!job->flags.bits.special && follow_job(RNG)) {
+            job_watched = job;
+            df::unit* unit = Job::getWorker(job);
+            if (df::global::ui && unit) {
+                our_dorf = unit;
+                df::global::ui->follow_unit = unit->id;
+            }
+        } else {
+            timestamp = tick;
+            std::vector<df::unit*> nonworkers;
+            for (auto unit: df::global::world->units.active) {
+                if (!Units::isCitizen(unit) || unit->job.current_job) {
+                    continue;
+                }
+                nonworkers.push_back(unit);
+            }
+            std::uniform_int_distribution<> follow_drunk(0, nonworkers.size() - 1);
+            if (df::global::ui) {
```

Comment:
no need to check for ui

Reply:
I have gone to remove these checks in the past and ended up with segfaults. I can try removing them again.. but I don't feel safe doing it